pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -171,46 +163,121 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
+			caBundle, err = ioutil.ReadFile(path.Join(LocalCertDir.Get(), "root-cert.pem"))
+			if err != nil {
+				return fmt.Errorf("failed reading %s: %v", path.Join(LocalCertDir.Get(), "root-cert.pem"), err)
+			}
 		}
-
 	} else {
 		log.Infof("User specified cert provider: %v", features.PilotCertProvider.Get())
 		return nil
 	}
-	if err != nil {
-		return err
-	}
+	s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
+	return nil
+}
 
-	// Save the certificates to ./var/run/secrets/istio-dns - this is needed since most of the code we currently
-	// use to start grpc and webhooks is based on files. This is a memory-mounted dir.
-	if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsKeyFile, keyPEM, 0o600)
-	if err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsCertFile, certChain, 0o600)
-	if err != nil {
-		return err
+// initCertificateWatches sets up watches for the dns certs.
+func (s *Server) initCertificateWatches(tlsOptions TLSOptions) error {
+	if hasCustomTLSCerts(tlsOptions) {
+		if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+			return fmt.Errorf("set keyCertBundle failed: %v", err)
+		}
+		// TODO: Setup watcher for root and restart server if it changes.
+		for _, file := range []string{tlsOptions.CertFile, tlsOptions.KeyFile} {
+			log.Infof("adding watcher for certificate %s", file)
+			if err := s.fileWatcher.Add(file); err != nil {
+				return fmt.Errorf("could not watch %v: %v", file, err)
+			}
+		}
+		s.addStartFunc(func(stop <-chan struct{}) error {
+			go func() {
+				var keyCertTimerC <-chan time.Time
+				for {
+					select {
+					case <-keyCertTimerC:
+						keyCertTimerC = nil
+						if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+							log.Errorf("Setting keyCertBundle failed: %v", err)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.CertFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.KeyFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case err := <-s.fileWatcher.Errors(tlsOptions.CertFile):
+						log.Errorf("error watching %v: %v", tlsOptions.CertFile, err)
+					case err := <-s.fileWatcher.Errors(tlsOptions.KeyFile):
+						log.Errorf("error watching %v: %v", tlsOptions.KeyFile, err)
+					case <-stop:
+						return
+					}
+				}
+			}()
+			return nil
+		})
+	} else if !s.EnableCA() && features.PilotCertProvider.Get() == IstiodCAProvider {
+		// no dns certs at all, this can only happen in test.
+		return nil
 	}
-	log.Info("DNS certificates created in ", dnsCertDir)
+
+	notifyCh := make(chan struct{})
+	watchCh := s.keyCertBundleWatcher.AddWatcher()
+	go s.reloadCert(watchCh, notifyCh)
```

Comment:
nit: this goroutine leaks and will never exit

Reply:
Will add a stop channel